src/background.ts

```diff
@@ -18,160 +17,196 @@ MainAccount m/84'/1776'/0'
-                index++;
-                const history = await getHistory(ws, addr.address.confidentialAddress);
-                if (history.length > 0) {
-                    count = 0;
-                    restoredIndex = index;
-                } else {
-                    count++;
-                }
-            }
-
-            // regenerate the address
-            for (let i = 0; i < restoredIndex; i++) {
-                await identity.getNextChangeAddress();
-            }
-            console.timeEnd('restore');
-
-            const addresses = await identity.getAddresses();
-
-            const keys: Record<string, AddressInterface> = {}
-            for (const addr of addresses) {
-                keys[AddressInfosKey(hashScriptAndReverse(addr.confidentialAddress))] = addr;
-            }
-            await chrome.storage.local.set(keys);
-
-            resolve(addresses);
-        });
+  const network = networks.testnet;
+
+  if (request.message === 'start_restore') {
+    console.log('start restore');
+
+    return new Promise(async (resolve, reject) => {
+      const mnemonic = request.mnemonic;
+
+      const seed = bip39.mnemonicToSeedSync(mnemonic);
+      const node = bip32.fromSeed(seed, network);
+    
+      const account = new Account(node, network);
+      const electrum = new ElectrumWS(
+        new WebSocket('wss://blockstream.info/liquidtestnet/electrum-websocket/api')
+      );
+
+      const { lastUsed, historyTxsId, heightsSet, txidHeight } = await sync(account, electrum);
+  
+      chrome.storage.local.set({ lastUsed, historyTxsId, heightsSet, txidHeight });
+      resolve({ lastUsed, historyTxsId, heightsSet, txidHeight });
+    });
+  }
+  return true;
+});
+
+
+
+async function sync(account: Account, electrum: ElectrumWS): Promise<{ 
+  lastUsed: { internal: number, external: number }, 
+  historyTxsId: Set<string>, 
+  heightsSet: Set<number>,
+  txidHeight: Map<string, number | undefined> 
+}> {
+
+  let historyTxsId: Set<string> = new Set();
+  let heightsSet: Set<number> = new Set();
+  let txidHeight: Map<string, number | undefined> = new Map();
+
+  let lastUsed = { internal: 0, external: 0 };
+  const walletChains = [0, 1];
+  for (const i of walletChains) {
+    const isInternal = i === 1;
+    let batchCount = 0;
+
+    while (true) {
+      const batch = account.deriveBatch(batchCount, GAP_LIMIT, isInternal);
+      const histories = await electrum.batchScriptGetHistory(batch);
+
+      let max = histories
+        .map((v, i) => v.length > 0 ? i : -1)
+        .reduce((a, b) => Math.max(a, b));
+      if (max >= 0) {
+        if (isInternal) {
+          lastUsed.internal = max + batchCount * GAP_LIMIT;
+        } else {
+          lastUsed.external = max + batchCount * GAP_LIMIT;
+        }
+      }
+
+
+      let flattened: GetHistoryResponse[] = histories.flat();
+      console.log(`${i}/batch(${batchCount}) ${flattened.length}`);
+
+      if (flattened.length === 0) {
```

Comment:
this is interesting: we stop looping when the flattened arrays contains 0 elements ie. empty gap limit